twilio-audio-router/src/main/java/com/twilio/audio_manager/AudioDeviceSelector.java

```diff
@@ -0,0 +1,376 @@
+package com.twilio.audio_manager;
+
+import android.bluetooth.BluetoothDevice;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.content.pm.PackageManager;
+import android.media.AudioAttributes;
+import android.media.AudioDeviceInfo;
+import android.media.AudioFocusRequest;
+import android.media.AudioManager;
+import android.os.Build;
+import android.os.Handler;
+import android.util.Log;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class AudioDeviceSelector {
+    private static final String TAG = "AudioDeviceSelector";
+
+    private final Context context;
+    private final AudioManager audioManager;
+    private final boolean hasEarpiece;
+    private final boolean hasSpeakerphone;
+    private final BluetoothController bluetoothController;
+
+    private @Nullable AudioDeviceChangeListener audioDeviceChangeListener;
+    private @Nullable AudioDevice selectedDevice;
+    private @Nullable AudioDevice userSelectedDevice;
+    private @NonNull State state;
+    private @NonNull WiredHeadsetReceiver wiredHeadsetReceiver;
+    private boolean wiredHeadsetAvailable;
+    private ArrayList<AudioDevice> availableAudioDevices = new ArrayList<>();
+
+    // Saved Audio Settings
+    private int savedAudioMode;
+    private boolean savedIsMicrophoneMuted;
+    private boolean savedSpeakerphoneEnabled;
+
+    private enum State {
+        STARTED,
+        ACTIVE,
+        STOPPED
+    }
+
+    private AudioDevice EARPIECE_AUDIO_DEVICE =
+            new AudioDevice(AudioDevice.Type.EARPIECE, "Earpiece");
+    private AudioDevice SPEAKERPHONE_AUDIO_DEVICE =
+            new AudioDevice(AudioDevice.Type.SPEAKERPHONE, "Speakerphone");
+    private AudioDevice WIRED_HEADSET_AUDIO_DEVICE =
+            new AudioDevice(AudioDevice.Type.WIRED_HEADSET, "Wired Headset");
+    private @Nullable AudioDevice bluetoothAudioDevice;
+
+    public AudioDeviceSelector(@NonNull Context context) {
+        ThreadUtils.checkIsOnMainThread();
```

Comment:
@idelgado is there a downstream component that requires all operations on the main thread? I'm not opposed to making this class single threaded, but it does seem strict to only allow the object itself to be created and operated on from only the main thread.

Reply:
Either way @Alton09 , we can sit down and work through the threading contract and document in the class level API documentation.

Reply:
Fair. My intent was to make it single threaded but I think it's possible to allow this library to be used on any thread rather than the main thread.

Reply:
I would cut a ticket for this and leave it as-is for now.